### PR TITLE
Document a training loop for streaming dataset

### PR DESCRIPTION
**Pull request by lhoestq**
I added some docs about streaming dataset. In particular I added two subsections:
- one on how to use `map` for preprocessing
- one on how to use a streaming dataset in a pytorch training loop

cc @patrickvonplaten @stevhliu if you have some comments
cc @Rocketknight1 later we can add the one for TF and I might need your help ^^'

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Looks great!

File: docs/source/stream.rst [hunk 1]
```diff
@@ -117,3 +117,70 @@ Remove columns on-the-fly with :func:`datasets.IterableDataset.remove_columns`.
    >>> from datasets import load_dataset
    >>> dataset = load_dataset('m4', 'en', streaming=True, split='train')
    >>> dataset = dataset.remove_columns('timestamp')
+
+``Map``
+^^^^^^^
+
+As for regular map-style :class:`datasets.Dataset`, 🤗 Datasets features :func:`datasets.IterableDataset.map` to process your dataset.
```
Comment (stevhliu) [inline docs/source/stream.rst:124]:
```suggestion
Similar to the :func:`datasets.Dataset.map` function for a regular :class:`datasets.Dataset`, 🤗  Datasets features :func:`datasets.IterableDataset.map` for processing :class:`datasets.IterableDataset`s.
```

File: docs/source/stream.rst [hunk 2]
```diff
@@ -117,3 +117,70 @@ Remove columns on-the-fly with :func:`datasets.IterableDataset.remove_columns`.
...
+Unlike a map-style dataset though, the processing is applied on-the-fly when the examples are streamed.
```
Comment (stevhliu) [inline docs/source/stream.rst:125]:
```suggestion
:func:`datasets.IterableDataset.map` applies processing on-the-fly when examples are streamed.
```

File: docs/source/stream.rst [hunk 3]
```diff
@@ -117,3 +117,70 @@ Remove columns on-the-fly with :func:`datasets.IterableDataset.remove_columns`.
...
+
+It allows you to apply a processing function to each example in a dataset, independently or in batches. This function can even create new rows and columns.
+
+In the following example, you will apply tokenization to the dataset. The function needs to accept and output a ``dict``:
```
Comment (stevhliu) [inline docs/source/stream.rst:129]:
```suggestion
The following example demonstrates how to tokenize a :class:`datasets.IterableDataset`. The function needs to accept and output a ``dict``:
```

File: docs/source/stream.rst [hunk 4]
```diff
@@ -117,3 +117,70 @@ Remove columns on-the-fly with :func:`datasets.IterableDataset.remove_columns`.
...
+   >>> tokenizer = AutoTokenizer.from_pretrained('distilbert-base-uncased')
+   >>> def encode(examples):
+   ...     return tokenizer(examples['text'], truncation=True, padding='max_length')
+   >>> dataset = dataset.map(encode, batched=True)
+   >>> next(iter(dataset))
+   {'input_ids': 101, 8466, 1018, 1010, 4029, 2475, 2062, 18558, 3100, 2061, ...,1106, 3739, 102],
+   'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ..., 1, 1]}
+
+
+In a training loop
```
Comment (stevhliu) [inline docs/source/stream.rst:146]:
```suggestion
Stream in a training loop
```

File: docs/source/stream.rst [hunk 5]
```diff
@@ -117,3 +117,70 @@ Remove columns on-the-fly with :func:`datasets.IterableDataset.remove_columns`.
...
+^^^^^^^^^^^^^^^^^^
+
+First let's shuffle the dataset:
```
Comment (stevhliu) [inline docs/source/stream.rst:149]:
```suggestion
:class:`datasets.IterableDataset`s can be integrated into a training loop. First, shuffle the dataset:
```

**Review by stevhliu — COMMENTED** [review]
Nice! Left a few minor comments to clarify map-style/streaming datasets :)

**Pull request merged**